python/ray/experimental/channel/cpu_nccl_group.py

```diff
@@ -0,0 +1,157 @@
+        # Buffer for the number of actors seen, each entry is one p2p op.
+        self.num_actors_seen = defaultdict(int)
+
+    async def wait(self, op_id: int, data=None):
+        """
+        Wait at barrier until all actors have sent `op_id`. One actor should
+        provide `data`, and this value will be returned by this method for all
+        other actors.
+        """
+        async with self.condition:
+            if data is not None:
+                assert op_id not in self.data, (self.data, self.num_actors_seen)
+                self.data[op_id] = data
+            self.num_actors_seen[op_id] += 1
+
+            if self.num_actors_seen[op_id] == self.num_actors:
+                # Wake up all tasks waiting on this condition.
+                self.condition.notify_all()
+            else:
+                await self.condition.wait_for(
+                    lambda: self.num_actors_seen[op_id] == self.num_actors
+                )
+
+            if data is None:
+                data = self.data[op_id]
+
+        return data
+
+    async def wait_collective(self, op_id: int, data: torch.Tensor, op: ReduceOp):
+        """
+        Wait at the barrier until all actors have sent `op_id` and `data`.
+        Once data from all actors is received, execute the collective `op`
+        on the barrier actor and return the result.
+        """
+        async with self.condition:
+            if op_id not in self.collective_data:
+                self.collective_data[op_id].append(data)
+            self.num_actors_seen[op_id] += 1
+
+            if self.num_actors_seen[op_id] == self.num_actors:
+                # Apply the collective operation across all gathered tensors
+                result = self._apply_op(op, self.collective_data[op_id])
+                self.collective_data[op_id] = result
+                self.condition.notify_all()
+            else:
+                await self.condition.wait_for(lambda: self.num_actors_seen[op_id] == self.num_actors)
+
+            # Return the result to all actors
+            return self.collective_data[op_id]
+
+    def _apply_op(self, op: ReduceOp, tensors: List[torch.Tensor]) -> torch.Tensor:
+        """Apply the specified reduction operation across a list of tensors."""
+        result = tensors[0].clone()
+        if op == ReduceOp.SUM:
+            for tensor in tensors[1:]:
+                result += tensor
+        elif op == ReduceOp.PRODUCT:
+            for tensor in tensors[1:]:
+                result *= tensor
+        elif op == ReduceOp.MAX:
+            for tensor in tensors[1:]:
+                result = torch.max(result, tensor)
+        elif op == ReduceOp.MIN:
+            for tensor in tensors[1:]:
+                result = torch.min(result, tensor)
+        elif op == ReduceOp.AVG:
+            result = sum(tensors) / len(tensors)
+        else:
+            # reserve a place for future ops to be added
+            assert False, "current operation not supported"
```

Comment:
Raise ValueError (assertion errors are for system-fatal bugs, not for errors that could happen because of bad user inputs).

Reply:
Fixed